cpp/src/arrow/filesystem/s3fs.h

```diff
@@ -69,6 +69,22 @@ enum class S3CredentialsKind : int8_t {
   WebIdentity
 };
 
+/// Pure virtual class for describing custom S3 retry strategies
+class S3RetryStrategy {
+ public:
+  struct AWSErrorDetail {
```

Comment:
nit: let's document that each of these fields corresponds to a field in the AWS SDK (even if that's rather trivial).